### PR TITLE
[#317] Content assist in OpenAPI3 doesn't show responses in project s…

### LINKED ISSUE #317: Content assist in OpenAPI3 doesn't show responses in project scope.
In this project:

[Multi-file Uber (OAS3).zip](https://github.com/RepreZen/KaiZen-OpenAPI-Editor/files/1023802/Multi-file.Uber.OAS3.zip)

Invoking code assist in `components/uber_path_user.yaml`, cycling to project scope, should suggest `"./uber_responses.yaml#/components/responses/ErrorResponse"`.  I can type in that path directly, and it passes validation without a warning.  

image

But it's not suggested in code assist:

image

### PR DESCRIPTION
**Pull request by ghillairet**
…cope.

Fixes #317 and #260

### CODE REVIEW

File: com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/JsonReferenceContextTest.xtend [hunk 1]
```diff
@@ -17,6 +17,16 @@ import static org.junit.Assert.*
 class JsonReferenceContextTest {
 
 	@Test
+	def void responses() {
+		assertTrue("/paths/~1foo/get/responses/200/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
+		assertTrue("/paths/~1foo/get/responses/2XX/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
+		assertTrue("/paths/~1foo/get/responses/3XX/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
+		assertTrue("/paths/~1foo/get/responses/default/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
+		assertFalse("/paths/~1foo/get/responses/2xx/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
+		assertFalse("/paths/~1foo/get/responses/3yz/$ref".matches(OpenApi3ReferenceProposalProvider.RESPONSE_REGEX))
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3.tests/src/com/reprezen/swagedit/openapi3/assist/JsonReferenceContextTest.xtend]: Thanks!

File: com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ReferenceProposalProvider.java [hunk 2]
```diff
@@ -29,7 +29,7 @@ public OpenApi3ReferenceProposalProvider() {
             // e.g. "/components/callbacks/myWebhook/$request.body#~1url/$ref"
             + "|/components/callbacks/" + COMPONENT_NAME_REGEX + "/[^/]+/\\$ref"; // in callbacks
     protected static final String PARAMETER_REGEX = ".*/parameters/\\d+/\\$ref";
-    protected static final String RESPONSE_REGEX = ".*/responses/\\d+/\\$ref";
+    protected static final String RESPONSE_REGEX = ".*/responses/((\\d{3}|\\dXX)|default)/\\$ref";
```
Comment (tfesenko) [inline com.reprezen.swagedit.openapi3/src/com/reprezen/swagedit/openapi3/assist/OpenApi3ReferenceProposalProvider.java]: This does no support response codes with a single wildcard, e.g. ` 20X`. 
What don't we use the regex from the schema: `"([0-9X]{3})$` (+ `|default`) ?

File: com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerReferenceProposalProvider.java [hunk 3]
```diff
@@ -24,7 +24,7 @@ public SwaggerReferenceProposalProvider() {
 	}
 
 	protected static final String SCHEMA_DEFINITION_REGEX = "^/definitions/(\\w+/)+\\$ref|.*schema/(\\w+/)?\\$ref";
-	protected static final String RESPONSE_REGEX = ".*responses/(\\d+|default)/\\$ref";
+    protected static final String RESPONSE_REGEX = ".*responses/(\\d{3}|default)/\\$ref";
```
Comment (tfesenko) [inline com.reprezen.swagedit/src/com/reprezen/swagedit/assist/SwaggerReferenceProposalProvider.java:27]: Nice

**Review by tfesenko — APPROVED** [review]
Looks good, but we need to support the `20X` and `2X1` cases . We can also support the `XXX` case as JSON Schema allows it.

**Review by tfesenko — CHANGES_REQUESTED** [review]
I actually wanted to request changes in my previous commit, but accidentally selected "Approve"

**Comment by ghillairet** [comment]
I changed the regex to support 2XX, 20X, XXX etc... But only for v3, should v2 supports it? It doesn't look like it is supported by validation.

**Comment by tfesenko** [comment]
> I changed the regex to support 2XX, 20X, XXX etc... 

Looks good, @ghillairet !
> But only for v3, should v2 supports it? It doesn't look like it is supported by validation.

v2 should only support three-digit combination + `default`

**Review by tfesenko — APPROVED** [review]

**Pull request merged**